### PR TITLE
zml/tensor: respect tags in broad

### LINKED ISSUE #68: `Tensor.broad` is too permissive
`.{ .a = 8, .l = 1, .k = 16}.broad(.{.a = 8, .k = 16, .v = 16})`; currently "compiles" but assign the `.l` on the left to `.k` and `.k` to `.v`. This should not be authorized when both left and right are tagged.

### PR DESCRIPTION
**Pull request by ousamabenyounes**
## Summary

Make `Tensor.broad` respect tags when both input and output shapes are fully tagged, and make `Shape.canBroadcastTo` reject fully tagged positional matches that do not map by tag.

## Issues

Fix #68

## Why

The issue case `.{ .a = 8, .l = 1, .k = 16 }.broad(.{ .a = 8, .k = 16, .v = 16 })` was accepted by positional broadcasting even though the tags do not describe the same axes. Fully tagged shapes should match and broadcast by tag, not by position.

The implementation keeps the integer scalar broadcast compatibility hook ahead of tagged broadcasting and excludes rank-0 shapes from the tagged branch so backend compatibility behavior is preserved.

## Test verification (RED -> GREEN)

Command: `./bazel.sh test //zml:test`

RED with the production fix reverted while keeping the new tests:

```text
53/142 shape.Shape.test.canBroadcastTo rejects fully tagged positional mismatch...FAIL (TestUnexpectedResult)
114/142 tensor.Tensor.test.Tensor.broad maps fully tagged axes by tag...thread ... panic: Can't broadcast Tensor({a=1,k=16,f32}) to {k=16,a=8,f32}
//zml:test                                                               FAILED in 23.5s
Executed 1 out of 1 test: 1 fails locally.
```

GREEN on the final branch:

```text
INFO: Build completed successfully, 6 total actions
//zml:test                                                               PASSED in 7.5s
Executed 1 out of 1 test: 1 test passes.
```

### CODE REVIEW

**Comment by gwenzek** [comment]
thanks, this is big change since it can break existing code. I'll try to re-prioritize this.

**Comment by ousamabenyounes** [comment]
I audited `.broad(` call sites to get a repo-level view of the impact of the stricter fully-tagged behavior.

Command used:

```bash
rg -n --glob '!**/*test*' --glob '!zml/tensor.zig' '\.broad\(' .
```

This returns 79 grep lines. Some lines contain more than one `.broad()` call, but I kept the audit at the grep-line level so it is directly reproducible.

Relevant implementation points:
- `Shape.canBroadcastTo()` now dispatches fully-tagged shapes to `canBroadcastToTagged()`.
- `Tensor.broad()` now uses tag axes when `self_shape.rank() != 0 && self_shape.isFullyTagged() && other.isFullyTagged()`.
- Therefore the risky pattern is: source shape fully tagged, target shape fully tagged, and at least one source tag is absent from the target.

Summary:
- `39` NOT_AFFECTED
- `35` SAFE_MATCHING
- `1` BROKEN
- `4` GENERIC_UNKNOWN

The only concrete broken call site I found is:

```zig
// examples/llm/models/qwen3_5_moe/model.zig:758
const shared_gate = self.shared_expert_gate.forward(x).sigmoid().broad(x.shape());
```

Reasoning:
- `shared_expert_gate` is initialized with a Linear weight shaped `{ .dout, .d }`.
- `zml.nn.Linear.forward()` calls `x.dot(self.weight, self.tag)`.
- Here the contracted tag is `.d`, so the dot result keeps the non-contracting weight tag `.dout`.
- Therefore `shared_expert_gate.forward(x)` has shape `{ .b, .s, .dout }`.
- `x.shape()` has shape `{ .b, .s, .d }`.
- With fully-tagged broadcast, `.dout` is absent from the target, so the new tagged path rejects this. The old positional path could allow it if the `.dout` dimension was 1.

Likely explicit fix:

```zig
const shared_gate = self.shared_expert_gate.forward(x).sigmoid().rename(.{ .dout = .d }).broad(x.shape());
```

Additional checks:
- The Llama/Qwen `token_index` broadcasts are scalar in the compiled/session paths (`Tensor.init(.{}, .u32)` / `Buffer.scalar`), so I classified those as NOT_AFFECTED rather than GENERIC_UNKNOWN.
- `zml/ops.zig:1366` is generic helper code for Neuron scatter/drop compatibility. The only repo caller I found that activates `ScatterOpts.increment` is `zml/moe/mosaic_tpu.zig:132` via `histogramCounts`; there, `ones` is created with `ids.shape()`, so the concrete repo path has matching shapes.

Validation:
- `./bazel.sh test //zml:test` passes locally.
- I could not complete `./bazel.sh test //...` locally because the full Bazel build exhausts available disk during dependency/toolchain analysis.

<details>
<summary>Full grep-line audit table</summary>

| Call site | Category | Notes |
|---|---:|---|
| examples/sharding/main.zig:36 | SAFE_MATCHING | bias axis matches output axis |
| zml/ops.zig:953 | GENERIC_UNKNOWN | scatter normalizes caller-provided index tensors to highest-rank index shape |
| zml/ops.zig:1115 | NOT_AFFECTED | constant/rank-0 source |
| zml/ops.zig:1216 | GENERIC_UNKNOWN | gather normalizes caller-provided index tensors to highest-rank index shape |
| zml/ops.zig:1366 | GENERIC_UNKNOWN | generic scatter/drop mask; only concrete repo increment caller checked safe |
| zml/ops.zig:1367 | NOT_AFFECTED | constant/rank-0 source |
| zml/ops.zig:1388 | NOT_AFFECTED | scalar source |
| zml/ops.zig:1389 | NOT_AFFECTED | scalar source |
| zml/ops.zig:1393 | NOT_AFFECTED | scalar source |
| zml/ops.zig:1397 | NOT_AFFECTED | constant/rank-0 source |
| zml/ops.zig:1417 | NOT_AFFECTED | constant/rank-0 source |
| zml/ops.zig:2300 | NOT_AFFECTED | constant/rank-0 source |
| examples/llm/models/llama/model.zig:409 | SAFE_MATCHING | `.d` weight broadcasts to x shape containing `.d` |
| examples/llm/models/llama/model.zig:518 | SAFE_MATCHING | `.s` to `.s` |
| examples/llm/models/llama/model.zig:519 | NOT_AFFECTED | `token_index` is scalar in inference/session paths |
| examples/llm/models/qwen3_5/model.zig:585 | SAFE_MATCHING | constructed `{ .b, .s }` to `{ .b, .s }` |
| examples/llm/models/qwen3_5/model.zig:586 | NOT_AFFECTED | `token_index` is scalar in inference/session paths |
| examples/llm/models/qwen3_5/model.zig:689 | SAFE_MATCHING | rope cos tags are inserted to match `x_rot.shape()` |
| examples/llm/models/qwen3_5/model.zig:690 | SAFE_MATCHING | same as previous |
| examples/llm/models/qwen3_5/model.zig:779 | NOT_AFFECTED | constant/rank-0 source |
| examples/llm/models/qwen3_5/model.zig:808 | NOT_AFFECTED | constant/rank-0 source |
| examples/llm/models/qwen3_5/model.zig:870 | SAFE_MATCHING | `aLog` / `dt_bias` axes are present in `a.shape()` |
| examples/llm/models/qwen3_5/model.zig:929 | SAFE_MATCHING | weight `.d` broadcasts to x shape containing `.d` |
| examples/llm/models/qwen3_5/model.zig:950 | SAFE_MATCHING | weight `.d` broadcasts to x shape containing `.d` |
| examples/llm/models/qwen3_5/model.zig:1016 | NOT_AFFECTED | cache layer index is scalar |
| zml/attention/nki/attention.zig:53 | GENERIC_UNKNOWN | public NKI attention helper; `token_index` shape comes from caller |
| zml/attention/flashattn.zig:991 | NOT_AFFECTED | constant/rank-0 source |
| zml/attention/flashattn.zig:1071 | NOT_AFFECTED | constant/rank-0 source |
| zml/attention/flashattn.zig:1132 | NOT_AFFECTED | constant/rank-0 source |
| zml/attention/flashattn.zig:1676 | NOT_AFFECTED | constant/rank-0 source |
| zml/attention/flashattn.zig:1744 | NOT_AFFECTED | constant/rank-0 source |
| zml/attention/flashattn.zig:1800 | NOT_AFFECTED | constant/rank-0 source |
| examples/llm/models/qwen3_5_moe/model.zig:587 | SAFE_MATCHING | constructed `{ .b, .s }` to `{ .b, .s }` |
| examples/llm/models/qwen3_5_moe/model.zig:588 | NOT_AFFECTED | `token_index` is scalar in inference/session paths |
| examples/llm/models/qwen3_5_moe/model.zig:758 | BROKEN | source `{ .b, .s, .dout }`, target `{ .b, .s, .d }` |
| examples/llm/models/qwen3_5_moe/model.zig:843 | SAFE_MATCHING | rope cos tags inserted to match `x_rot.shape()` |
| examples/llm/models/qwen3_5_moe/model.zig:844 | SAFE_MATCHING | same as previous |
| examples/llm/models/qwen3_5_moe/model.zig:933 | NOT_AFFECTED | constant/rank-0 source |
| examples/llm/models/qwen3_5_moe/model.zig:961 | NOT_AFFECTED | constant/rank-0 source |
| examples/llm/models/qwen3_5_moe/model.zig:1035 | SAFE_MATCHING | `aLog` / `dt_bias` axes are present in `a.shape()` |
| examples/llm/models/qwen3_5_moe/model.zig:1042 | SAFE_MATCHING | `valid_mask` has `.s`, target contains `.s` |
| examples/llm/models/qwen3_5_moe/model.zig:1043 | SAFE_MATCHING | same as previous |
| examples/llm/models/qwen3_5_moe/model.zig:1044 | SAFE_MATCHING | same as previous |
| examples/llm/models/qwen3_5_moe/model.zig:1094 | SAFE_MATCHING | weight `.d` broadcasts to x shape containing `.d` |
| examples/llm/models/qwen3_5_moe/model.zig:1115 | SAFE_MATCHING | weight `.d` broadcasts to x shape containing `.d` |
| examples/llm/models/qwen3_5_moe/model.zig:1178 | NOT_AFFECTED | cache layer index is scalar |
| examples/llm/models/lfm2/model.zig:422 | SAFE_MATCHING | scalar `left_pad` plus same-shape `iota(sh).broad(sh)` |
| examples/llm/models/lfm2/model.zig:445 | SAFE_MATCHING | kernel tags `{ .seq, .d }` are present in target `{ .batch, .seq, .d }` |
| examples/llm/models/lfm2/model.zig:508 | SAFE_MATCHING | target is built by inserting `.seq` into `tokens_position_offset.shape()` |
| examples/llm/models/lfm2/model.zig:631 | NOT_AFFECTED | `cache_index` is scalar |
| examples/llm/models/lfm2/model.zig:659 | SAFE_MATCHING | Linear bias broadcasts to `y.shape()` containing output tag |
| examples/llm/models/lfm2/model.zig:702 | SAFE_MATCHING | norm weight tag is present in input shape |
| zml/nn.zig:32 | SAFE_MATCHING | Linear bias broadcasts to `y.shape()` containing output tag |
| zml/nn.zig:99 | SAFE_MATCHING | reduction result tags are a subset of `x.shape()` |
| zml/nn.zig:296 | SAFE_MATCHING | rope cos tags are subset/present in `x_real.shape()` |
| zml/nn.zig:297 | SAFE_MATCHING | same as previous |
| zml/nn.zig:1009 | NOT_AFFECTED | constant/rank-0 source |
| zml/nn.zig:1076 | NOT_AFFECTED | constant/rank-0 source |
| zml/nn.zig:1092 | NOT_AFFECTED | constant/rank-0 source |
| zml/nn.zig:1160 | NOT_AFFECTED | constant/rank-0 source |
| zml/nn.zig:1161 | NOT_AFFECTED | constant/rank-0 source |
| zml/nn.zig:1222 | SAFE_MATCHING | `sdpa` validates/collects `.h/.q/.k/.hd`; mask tags are expected subset of attention weights |
| zml/nn.zig:1303 | SAFE_MATCHING | inserted `.v/.k` tags match `state.s.shape()` |
| zml/nn.zig:1679 | NOT_AFFECTED | DynamicSamplingStrategy tensors are scalar buffers |
| zml/nn.zig:1681 | NOT_AFFECTED | same as previous |
| zml/nn.zig:1690 | NOT_AFFECTED | same as previous |
| zml/nn.zig:1691 | SAFE_MATCHING | scalar to `probs_max`, then `.topk` subset to `x.shape()` |
| zml/attention/tpu_attention.zig:116 | NOT_AFFECTED | constant/rank-0 source |
| zml/moe/mosaic_tpu.zig:131 | NOT_AFFECTED | constant/rank-0 source |
| zml/moe/mosaic_tpu.zig:193 | SAFE_MATCHING | inserted `.token_src` matches `onehot_shape` |
| zml/moe/mosaic_tpu.zig:197 | SAFE_MATCHING | inserted `.route`/`.token_src` match `onehot_shape` |
| zml/moe/mosaic_tpu.zig:277 | NOT_AFFECTED | scalar source |
| zml/moe/mosaic_tpu.zig:279 | NOT_AFFECTED | scalar source |
| zml/moe/mosaic_tpu.zig:280 | NOT_AFFECTED | scalar source |
| zml/moe/mosaic_tpu.zig:318 | NOT_AFFECTED | scalar source |
| zml/moe/mosaic_tpu.zig:323 | NOT_AFFECTED | scalar source |
| zml/moe/mosaic_tpu.zig:367 | SAFE_MATCHING | router weights route/token tag is present in down output |
| zml/moe/metal.zig:112 | SAFE_MATCHING | inserted `.top_expert` matches target `{ .token, .top_expert, .d }` |
| zml/moe/metal.zig:136 | SAFE_MATCHING | weights route tag is present in down output |

</details>
